src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,355 @@
+        {
+            BoundExpression hasValueCall = MakeNullableHasValue(expression.Syntax, expression);
+            return MakeTemp(hasValueCall, temps, effects);
+        }
+
+        /// <summary>
+        /// Returns a temp which is initialized with lowered-expression.GetValueOrDefault()
+        /// </summary>
+        private BoundLocal MakeValueOrDefaultTemp(BoundExpression expression,
+            ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> effects)
+        {
+            BoundExpression valueOrDefaultCall = MakeOptimizedGetValueOrDefault(expression.Syntax, expression);
+            return MakeTemp(valueOrDefaultCall, temps, effects);
+        }
+
+        /// <summary>
+        /// Produces a chain of equality (or inequality) checks combined logically with AND (or OR)
+        /// </summary>
+        private BoundExpression RewriteNonNullableNestedTupleOperators(TupleBinaryOperatorInfo.Multiple operators,
+            BoundExpression left, BoundExpression right, TypeSymbol type,
+            ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> effects, TupleOperatorSideEffectsAndTemps effectAndTemps, BinaryOperatorKind operatorKind)
+        {
+            ImmutableArray<TupleBinaryOperatorInfo> nestedOperators = operators.Operators;
+
+            BoundExpression currentResult = null;
+            for (int i = 0; i < nestedOperators.Length; i++)
+            {
+                BoundExpression leftElement = GetTuplePart(left, i, effectAndTemps.temps, effectAndTemps.leftInit);
+                BoundExpression rightElement = GetTuplePart(right, i, effectAndTemps.temps, effectAndTemps.rightInit);
+                BoundExpression nextLogicalOperand = RewriteTupleOperator(nestedOperators[i], leftElement, rightElement, type, effectAndTemps, operatorKind);
+                if (currentResult is null)
+                {
+                    currentResult = nextLogicalOperand;
+                }
+                else
+                {
+                    var logicalOperator = operatorKind == BinaryOperatorKind.Equal ? BinaryOperatorKind.LogicalBoolAnd : BinaryOperatorKind.LogicalBoolOr;
+                    currentResult = _factory.Binary(logicalOperator, type, currentResult, nextLogicalOperand);
+                }
+            }
+
+            return currentResult;
+        }
+
+        /// <summary>
+        /// For tuple literals, we just return the element.
+        /// For expressions with tuple type, we access `Item{i}`.
+        /// </summary>
+        private BoundExpression GetTuplePart(BoundExpression tuple, int i, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> initEffects)
+        {
+            // Example:
+            // (1, 2) == (1, 2);
+            if (IsTupleExpression(tuple.Kind))
+            {
+                return ((BoundTupleExpression)tuple).Arguments[i];
```

Comment:
>return ((BoundTupleExpression)tuple).Arguments[i]; [](start = 16, length = 50)

Can this be anything, but a tuple literal?

Reply:
The implementation of `IsTupleExpression` is `return kind == BoundKind.TupleLiteral || kind == BoundKind.ConvertedTupleLiteral;` and the two sub-types of `BoundTupleExpression` are `BoundTupleLiteral` and `BoundConvertedTupleLiteral`.
  

---
In reply to: [172270998](https://github.com/dotnet/roslyn/pull/23356#discussion_r172270998) [](ancestors = 172270998)

Reply:
> The implementation of IsTupleExpression is return kind == BoundKind.TupleLiteral || kind == BoundKind.ConvertedTupleLiteral; and the two sub-types of BoundTupleExpression are BoundTupleLiteral and BoundConvertedTupleLiteral.

Given the behavior of ReplaceTerminalElementsWithTemps, it feels like we cannot get here for anything, but a tuple literal 

---
In reply to: [172275316](https://github.com/dotnet/roslyn/pull/23356#discussion_r172275316) [](ancestors = 172275316,172270998)